operator/pkg/ciliumendpointslice/endpointslice.go

```diff
@@ -48,263 +41,223 @@ const (
 		// only the overall factor (not per item).
-		&workqueue.BucketRateLimiter{Limiter: rate.NewLimiter(rate.Limit(qpsLimit), burstLimit)},
+		&workqueue.BucketRateLimiter{Limiter: rate.NewLimiter(rate.Limit(c.writeQPSLimit), c.writeQPSBurst)},
 	), "cilium_endpoint_slice")
+	c.queueTerminated = make(chan struct{})
+}
 
-	manager := newCESManagerFcfs(rlQueue, maxCEPsInCES)
-	if slicingMode == cesIdentityBasedSlicing {
-		manager = newCESManagerIdentity(rlQueue, maxCEPsInCES)
+func (c *Controller) onEndpointUpdate(cep *cilium_api_v2.CiliumEndpoint) {
+	if cep.Status.Networking == nil || cep.Status.Identity == nil || cep.GetName() == "" || cep.Namespace == "" {
+		return
 	}
-	cesStore := ciliumEndpointSliceInit(clientset.CiliumV2alpha1(), ctx, wg)
-	ceSliceStore = cesStore
-
-	// List all existing CESs from the api-server and cache it locally.
-	// This sync should happen before starting CEP watcher, because CEP watcher
-	// emits the existing CEPs as newly added CEPs. If we don't have local sync
-	// cesManager would assume those are new CEPs and may create new CESs for those CEPs.
-	// This situation ends up having duplicate CEPs in different CESs. Hence, we need
-	// to sync existing CESs before starting a CEP watcher.
-	syncCESsInLocalCache(cesStore, manager)
-	return &CiliumEndpointSliceController{
-		clientV2:                 clientset.CiliumV2(),
-		clientV2a1:               clientset.CiliumV2alpha1(),
-		reconciler:               newReconciler(clientset.CiliumV2alpha1(), manager),
-		Manager:                  manager,
-		queue:                    rlQueue,
-		ciliumEndpointSliceStore: cesStore,
-		slicingMode:              slicingMode,
-		workerLoopPeriod:         1 * time.Second,
+	touchedCESs := c.manager.UpdateCEPMapping(k8s.ConvertCEPToCoreCEP(cep), cep.Namespace)
+	c.enqueueCESReconciliation(touchedCESs)
+}
+
+func (c *Controller) onEndpointDelete(cep *cilium_api_v2.CiliumEndpoint) {
+	touchedCES := c.manager.RemoveCEPMapping(k8s.ConvertCEPToCoreCEP(cep), cep.Namespace)
+	c.enqueueCESReconciliation([]CESName{touchedCES})
+}
+
+func (c *Controller) onSliceUpdate(ces *capi_v2a1.CiliumEndpointSlice) {
+	c.enqueueCESReconciliation([]CESName{NewCESName(ces.Name)})
+}
+
+func (c *Controller) onSliceDelete(ces *capi_v2a1.CiliumEndpointSlice) {
+	c.enqueueCESReconciliation([]CESName{NewCESName(ces.Name)})
+}
+
+func (c *Controller) enqueueCESReconciliation(cess []CESName) {
+	for _, ces := range cess {
+		c.logger.WithFields(logrus.Fields{
+			logfields.CESName: ces.string(),
+		}).Debug("Enqueueing CES (if not empty name)")
+		if ces.Name != "" {
+			c.enqueuedAtLock.Lock()
+			if c.enqueuedAt[ces].IsZero() {
+				c.enqueuedAt[ces] = time.Now()
+			}
+			c.enqueuedAtLock.Unlock()
+			c.queue.AddAfter(ces, DefaultCESSyncTime)
+		}
+	}
+}
+
+func (c *Controller) getAndResetCESProcessingDelay(ces CESName) float64 {
+	enqueued, exists := c.enqueuedAt[ces]
+	if !exists {
+		return 0
+	}
+	c.enqueuedAtLock.Lock()
```

Comment:
I think we should take the lock before checking the key existence, too. That's because it is possible to have a concurrent write to the `c.enqueuedAt` map at line 109 and that can potentially reallocate the map internal buckets (e.g: think about when the map grows) giving us incorrect results.

Reply:
Done